Assets/Plugins/iOS/Shopify/UnityMessageSender.swift

```diff
@@ -0,0 +1,69 @@
+//
+//  UnityMessageSender.swift
+//  UnityBuySDK
+//
+//  Created by Shopify.
+//  Copyright © 2017 Shopify Inc. All rights reserved.
+//
+//  Permission is hereby granted, free of charge, to any person obtaining a copy
+//  of this software and associated documentation files (the "Software"), to deal
+//  in the Software without restriction, including without limitation the rights
+//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions:
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import Foundation
+import PassKit
+
+typealias MessageCompletion = (_ result: String?) -> Void
+
+@objc class UnityMessageSender: NSObject {
+
+    static var pendingMessageResponses = [String: UnityMessageResponse]()
+    
+    static func sendMessage(_ message: String, toObject objectName: String, havingMethod methodName: String) -> String? {
+        
+        let message  = UnityMessage.init(message: message);
+        let response = UnityMessageResponse.init(identifier: message.identifier, completion: DispatchSemaphore(value: 0))
```

Comment:
The fact that a `UnityMessageResponse` is always created within the same scope as a `UnityMessage` tells me that there's no benefit to having two separate objects represent these actions. To me, a `UnityMessage` is a single unit of work where we send a message, wait for the receiver and complete with an optional result. Splitting these up into two objects that share most of the state doesn't provide any benefit in this particular case. If anything, it adds complexity in keeping the two objects in sync. I think an ideal interface would look something like this:
```swift
// Somewhere at the call site
let message = Message(to: object, method: method, message: "do_something")
UnityMessageSender.send(message)

// inside UnityMessageSender
static func send(_ message: Message) {
    let identifier = message.identifier

    messages[identifier] = message
    UnityInterfaceWrapper.sendMessage(try! message.content(), toObject: message.object, havingMethodName: message.method)

    message.wait { result in
        messages[identifier] = nil
        completionHandler(result)
    }
}
```
The idea is that you're passing `Message` around instead of loose values like `message`, `objectName` and `methodName`. The `Message` should encapsulate those.